hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java

```diff
@@ -1285,7 +1285,9 @@ private boolean waitForNamespaceOnline() throws InterruptedException, IOExceptio
     }
     // Else there are namespace regions up in meta. Ensure they are assigned before we go on.
     for (RegionInfo ri : ris) {
-      isRegionOnline(ri);
+      if (!isRegionOnline(ri)) {
```

Comment:
Actually isRegionOnline() is waiting in a loop until this region's status become opened and that server is online.  So there is no bug as such right?  isRegionOnline() might return false iff the server is being stopped

Reply:
`isRegionOnline` might also return false if region doesn't come online after all retry attempts with exponential backoff are completed (rare but valid case)

Reply:
BTW our RetryCounterFactory is with infinite retry. maxAttempts are Integer.MAX_VALUE.  Also we don't account for a case whether RetryCounterFactory finishes the retry attempts. 
There is no harm in adding this fix.  But should we see whether the RetryCounterFactory  usage is correct also?
Am I missing some thing?

Reply:
Oh I see, for infinite retries, it is not a problem but then ideally we should change the signature of `isRegionOnline()` because boolean is of no use. Also, it is being used in similar manner for meta region also, and that's why I wanted to make namespace region open logic look same.

e.g
```
  public boolean waitForMetaOnline() throws InterruptedException {
    return isRegionOnline(RegionInfoBuilder.FIRST_META_REGIONINFO);
  }

```
However, now I understand the main purpose of `isRegionOnline()` which is to keep retrying infinite times and only when server is going down, return false. And with the current logic, `waitForNamespaceOnline()` returns true no matter what `isRegionOnline()` returns and hence, instead of returning from HMaster active initializations, it proceeds further.

```
    if (!waitForNamespaceOnline()) {
      return;
    }
    status.setStatus("Starting cluster schema service");
    initClusterSchemaService();
....
....
....
```
We should return from the above if condition ideally, but anyways, this is not significant problem given that only condition this happens, server will be shutting down. But good to maintain same logic for namespace and meta regions initializations. Thought?

Reply:
@ramkrish86 from this code block, we just return false and we return back from `finishActiveMasterInitialization()` without any further initialization because mostly this will happen when HM server is going down so we expect `finishActiveMasterInitialization()` to return instead of going further with any further init:

`finishActiveMasterInitialization()`:
```
    if (!waitForNamespaceOnline()) {
      return;
    }

```

`waitForNamespaceOnline()`:
```
      if (!isRegionOnline(ri)) {
        return false;
      }
```

Reply:
Ya seeing the meta case this seems fine. Just a unification. +1.